src/iter/walk_tree.rs

```diff
@@ -0,0 +1,545 @@
+    }
+}
+
+/// Create a tree-like prefix parallel iterator from an initial root node.
+/// The `children_of` function should take a node and return an iterator over its child nodes.
+/// The best parallelization is obtained when the tree is balanced
+/// but we should also be able to handle harder cases.
+///
+/// # Ordering
+///
+/// This function guarantees a prefix ordering. See also [`walk_tree_postfix`],
+/// which guarantees a postfix order.
+/// If you don't care about ordering, you should use [`walk_tree`],
+/// which will use whatever is believed to be fastest.
+/// Between siblings, children are reduced in reverse order -- that is, the children that returned last are reduced first.
+/// For example a perfect binary tree of 7 nodes will reduced in the following order:
+///
+/// ```text
+///      a
+///     / \
+///    /   \
+///   b     c
+///  / \   / \
+/// d   e f   g
+///
+/// reduced as  a,b,d,e,c,f,g
+///
+/// ```
+///
+///
+/// For a postfix ordering see the (faster) [`walk_tree_postfix()`] function.
+///
+/// [`walk_tree_postfix()`]: fn.walk_tree_postfix.html
+/// [`walk_tree()`]: fn.walk_tree.html
+///
+/// # Example
+///
+/// ```text
+///      4
+///     / \
+///    /   \
+///   2     3
+///        / \
+///       1   2
+/// ```
+///
+/// ```
+/// use rayon::prelude::*;
+/// use rayon::iter::walk_tree_prefix;
+/// assert_eq!(
+///     walk_tree_prefix(4, |&e| if e <= 2 { Vec::new() } else {vec![e/2, e/2+1]})
+///         .sum::<u32>(),
+///     12);
+/// ```
+///
+/// # Example
+///
+///    ```
+///    use rayon::prelude::*;
+///    use rayon::iter::walk_tree_prefix;
+///
+///    struct Node {
+///        content: u32,
+///        left: Option<Box<Node>>,
+///        right: Option<Box<Node>>,
+///    }
+///
+/// // Here we loop on the following tree:
+/// //
+/// //       10
+/// //      /  \
+/// //     /    \
+/// //    3     14
+/// //            \
+/// //             \
+/// //              18
+///
+///    let root = Node {
+///        content: 10,
+///        left: Some(Box::new(Node {
+///            content: 3,
+///            left: None,
+///            right: None,
+///        })),
+///        right: Some(Box::new(Node {
+///            content: 14,
+///            left: None,
+///            right: Some(Box::new(Node {
+///                content: 18,
+///                left: None,
+///                right: None,
+///            })),
+///        })),
+///    };
+///    let mut v: Vec<u32> = walk_tree_prefix(&root, |r| {
+///        r.left
+///            .as_ref()
+///            .into_iter()
+///            .chain(r.right.as_ref())
+///            .map(|n| &**n)
+///    })
+///    .map(|node| node.content)
+///    .collect();
+///    assert_eq!(v, vec![10, 3, 14, 18]);
+///    ```
+///
+pub fn walk_tree_prefix<S, B, I, IT>(root: S, breed: B) -> WalkTreePrefix<S, B, I>
+where
+    S: Send,
+    B: Fn(&S) -> I + Send + Sync,
+    IT: DoubleEndedIterator<Item = S>,
+    I: IntoIterator<Item = S, IntoIter = IT> + Send,
```

Comment:
I think we can remove `IT` as a public API parameter by constraining `I::IntoIter: Double...` instead.

Should we also require `I::IntoIter: Send`? We consume it immediately now, but might there be a future where we'd want to be more lazy about that?

Reply:
i'm not sure i get it. what would send allow us to do ? IT would not be a parallel iterator so it would just mean using another thread. `I` is already Send.
so i'm not getting what this would potentially allow us to do and why `I` being Send is not enough

Reply:
`I: Send` doesn't actually give us much, only that after getting such a return value of `B`, we could hold and send that `I` value to other threads *before* calling `into_iter()`. Once that's called, we now have the `IntoIter` type, which we might want to lazily pull items out one at a time.

For example, suppose `to_explore` were instead a `Vec<Either<S, IntoIter>>`, storing either unvisited nodes or their unprocessed (or partially processed) brood. We would need `IntoIter: Send` to be able to split this between threads.

Maybe we don't need to do this, especially if we don't expect there to be very many children most of the time.

Reply:
Actually, `I: Send` is only necessary right now because you have that `PhantomData<I>`, but I don't think you actually need that. The structs themselves don't need the `I` parameter at all, only the `impl` constraints.